### DISCUSSION TITLE
Create a model based on another model so it tracks changes on the Rust side

### DISCUSSION THREAD

**Discussion by kanashimia** [slint-ui/slint#5413] [Q&A]
```slint
export component Demo inherits Window {
    in-out property <[Node]> nodes: [
        { text: "Node1", x: 120px,  y: 235px, } ,
        { text: "Node2", x: 350px,  y: 150px, } ,
        { text: "Node3", x: 300px,  y: 350px, } ,
    ];

    function create-link(a:int, b:int)-> Link {
        return { x1: nodes[a].x, y1: nodes[a].y, x2: nodes[b].x, y2: nodes[b].y };
    }
    in property <[Link]> links: [ create-link(1,2), create-link(0,2) ];
}
```
Code from https://github.com/slint-ui/slint/discussions/3608#discussioncomment-7207488 again, node graph struggles.
Here `links` reacts to the changes of `nodes`

But when you do 
```rust
let links_model: Rc<VecModel<_>> = Rc::new(
    vec![
        main_window.invoke_create_link(0, 1),
        main_window.invoke_create_link(1, 2),
    ]
    .into(),
);
main_window.set_links(links_model.clone().into());
```

Then it obviously doesn't react because it just copies the value.
So how do I make Rust code behave like it is in Slint?

The only idea I have right now is to make a custom implementation of the Model, like in the 7guis/cells example, but I don't know how.
I've looked at the generated code, it uses `row_data_tracked` but I don't know how to use it and if it is useful at all.

Instead of this I can do
```slint
in-out property <[{ from: int, to: int }]> bindings: [{from: 1, to: 2}, {from: 0, to: 2}];
for b in bindings: path := Path {
    property <Link> l: create-link(b.from, b.to);
```
But am still interested the solution.

Another solution is to use workaround like https://github.com/slint-ui/slint/discussions/4717 and recreate every link. 

Honestly it is kinda strange that it reacts to model changes at all in slint, but whatever.

**Comment by Vadoola** [comment]
I havent looked into this at all, or created a custom model for Slint. I did see in the docs somewhere there is a model peer struxt (I think) that can be notified if another model changes. I imagine you would want that? Perhaps the source code for FilterModel would be a place to start looking?

**Comment by ogoffart** [comment]
I think the solution you suggest with the `property <Link> l: create-link(b.from, b.to);` is the best at the moment. 
The solution in pure Slint that use the create-link in the model expression is not really optimal as it will re-create the whole model when something changes.
